File: 2018/03-fabric-cutting/python/star1.py
#!/usr/bin/env python3

from collections import defaultdict

def get_squares(claim):
    id_, rest = claim[1:].split(' @ ')
    location, size = rest.split(': ')
    x, y = [int(i) for i in location.split(',')]
    width, height = [int(i) for i in size.split('x')]
    squares = []
    for i in range(x, x+width):
        for j in range(y, y+height):
            squares.append((i,j))
    return squares

with open('input') as f:
    claims = f.readlines()

    claimed_squares = set()
    overlapping_squares = set()
    for claim in claims:
        squares = get_squares(claim)
        for square in squares:
            if square in claimed_squares:
                overlapping_squares.add(square)
            claimed_squares.add(square)

    print(len(overlapping_squares))

File: 2018/03-fabric-cutting/go/star1.go
package main

import (
    "io/ioutil"
    "fmt"
    "strings"
)

func main() {
    bytes, _ := ioutil.ReadFile("input")
    lines := strings.Split(string(bytes), "\n")

    claimedSquares := make(map[[2]int]int)
    for _, line := range lines {
        if line == "" {
            continue
        }
        var id, x, y, w, h int
        fmt.Sscanf(line, "#%d @ %d,%d: %dx%d", &id, &x, &y, &w, &h)

        for i := x; i < x+w; i++ {
            for j := y; j < y+h; j++ {
                claimedSquares[[2]int{i, j}] += 1
            }
        }
    }

    doubleClaims := 0
    for _, claims := range claimedSquares {
        if claims > 1 {
            doubleClaims += 1
        }
    }

    fmt.Println(doubleClaims)
}
